src/components/selection-list/selection-list-model.tsx

```diff
@@ -1,166 +1,266 @@
+            }
+            if (typeof this.item.props.children === 'string') {
+                return this.item.props.children;
+            }
         }
+        return '';
+    }
+
+    public get disabled() {
+        return Boolean(typeof this.item === 'object' && this.item.props.disabled);
     }
 
-    public focusValue(value: SelectionListItemValue | undefined) {
-        this.focusedValue = value;
-        for (const item of this.items) {
-            item.focused = (value !== undefined && item.value === value);
+    public render(id: string) {
+        if (typeof this.item !== 'object') {
+            return null;
         }
+
+        return this.isOption ?
+            React.cloneElement(this.item, {id, selected: this.selected, focused: this.focused}) : this.item;
+    }
+}
+
+export function selectionListItemsFromDataSource(
+    dataSource: OptionList['dataSource'] = [],
+    resolveItem: OptionList['dataMapper'] = defaultResolveItem,
+    renderItem: OptionList['renderItem'] = defaultRenderItem
+): SelectionListItem[] {
+    return dataSource.map(item => new SelectionListItemBasedOnDataSource(item, resolveItem, renderItem));
+}
+
+export function selectionListItemsFromChildren(children: React.ReactNode): SelectionListItem[] {
+    return React.Children.map(children, item => new SelectionListItemBasedOnChild(item));
+}
+
+export function selectionListItemsFromProps(props: OptionList & {children?: React.ReactNode}): SelectionListItem[] {
+    return selectionListItemsFromChildren(props.children || []).concat(
+        selectionListItemsFromDataSource(props.dataSource, props.dataMapper, props.renderItem)
+    );
+}
+
+export class SelectionListModel {
+    @observable public items: SelectionListItem[] = [];
+    @observable public focusedIndex: number = -1;
+    @observable public selectedIndex: number = -1;
+
+    constructor(items: SelectionListItem[]) {
+        this.items = items;
+    }
+
+    @action public selectValue(value: SelectionListItemValue) {
+        this.selectIndex((value === undefined) ? -1 : this.items.findIndex(item => item.value === value));
     }
 
-    private addItem(data: SelectionListItemData, element: JSX.Element) {
-        const value = element.props.value;
-        const disabled = Boolean(element.props.disabled);
-        const item = extendShallowObservable(
-            {
-                data,
-                disabled,
-                element,
-                isOption: value !== undefined,
-                selectable: value !== undefined && !disabled,
-                value
-            }, {
-                focused: false,
-                selected: false
+    public hasSelection() {
+        return this.selectedIndex > -1;
+    }
+
+    @action public focusSelected() {
+        if (this.selectedIndex > -1 && this.items[this.selectedIndex].selectable) {
+            this.focusIndex(this.selectedIndex);
+        }
+    }
+
+    public typeAhead(character: string) {
```

Comment:
is there a reason to have this method available?

Reply:
Just as a reminder that this is where it should go.